clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1490,6 +1490,8 @@ public void unsubscribe() {
                     log.error("Failed while waiting for the unsubscribe event to complete");
                 }
                 resetGroupMetadata();
+            } else {
```

Comment:
inspired by @lianetm comments in jira. Maybe we can leverage `UnsubscribeEvent` to handle the `unsubscribe`. For example:

```java
    public void unsubscribe() {
        acquireAndEnsureOpen();
        try {
            fetchBuffer.retainAll(Collections.emptySet());
            Timer timer = time.timer(Long.MAX_VALUE);
            UnsubscribeEvent unsubscribeEvent = new UnsubscribeEvent(calculateDeadlineMs(timer));
            applicationEventHandler.add(unsubscribeEvent);
            log.info("Unsubscribing all topics or patterns and assigned partitions {}",
                    subscriptions.assignedPartitions());
            try {
                processBackgroundEvents(unsubscribeEvent.future(), timer);
                log.info("Unsubscribed all topics or patterns and assigned partitions");
            } catch (TimeoutException e) {
                log.error("Failed while waiting for the unsubscribe event to complete");
            }
            if (groupMetadata.get().isPresent()) resetGroupMetadata();
        } catch (Exception e) {
            log.error("Unsubscribe failed", e);
            throw e;
        } finally {
            release();
        }
    }

```

Reply:
yeap, I really liked that idea when I shared it. We should only note that if we take that path, we would have to end up calling the `subcriptionState.unsubscribe` from the `ApplicationEventProcessor` where we process(UnsubscribeEvent). We wouldn't be able to reuse the `MembershipMgr` logic (which is part of what I had in mind when commenting on the jira), because if there is no groupId we don't even have an instance of the MembershipMgr. 

Still, I like the idea because 1) simplifies the consumer.unsubscribe, 2) all subscriptionState.unsubscribe calls are in the background thread associated to the UnsubscribeEvent (directly if no groupId, indirectly if groupId present, delegating on the membershipMgr). Makes sense?

Reply:
I like the idea too. 

> we would have to end up calling the subcriptionState.unsubscribe from the ApplicationEventProcessor where we process(UnsubscribeEvent). We wouldn't be able to reuse the MembershipMgr logic (which is part of what I had in mind when commenting on the jira), because if there is no groupId we don't even have an instance of the MembershipMgr.

nice point!

Reply:
Hi @chia7712, thanks for the suggestion. I also updated related test case.

Reply:
Hi @chia7712 and @lianetm, it looks like we can't use `UnsubscribeEvent` when group id is empty, because `heartbeatRequestManager` will be empty. The test case `PlaintextConsumerTest#testConsumingWithNullGroupId` can't pass.

https://github.com/apache/kafka/blob/9a7eee60727dc73f09075e971ea35909d2245f19/clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java#L215-L220

https://github.com/apache/kafka/blob/9a7eee60727dc73f09075e971ea35909d2245f19/clients/src/main/java/org/apache/kafka/clients/consumer/internals/RequestManagers.java#L161-L208

Reply:
@FrankYang0529 Please take a look at @lianetm comments:

> we would have to end up calling the subcriptionState.unsubscribe from the ApplicationEventProcessor where we process(UnsubscribeEvent). We wouldn't be able to reuse the MembershipMgr logic (which is part of what I had in mind when commenting on the jira), because if there is no groupId we don't even have an instance of the MembershipMgr.

Reply:
Hi @chia7712, sorry, I didn't notice that. I introduce `SubscriptionState` to `ApplicationEventProcessor`, call `subscriptions.unsubscribe()` for `UnsubscribeEvent`, and remove `subscriptions.unsubscribe()` from `MembershipManagerImpl#leaveGroup`.